### PR TITLE
initial steps on complete data function

### PR DESCRIPTION
**Pull request by shannonpileggi**
still a WIP, not ready for review!

### CODE REVIEW

**Comment by shannonpileggi** [comment]
Hi @ddsjoberg! I have a first pass at the `complete_data()` function. I can write more tests & checks, but before I do, can you review to see if you agree? (I haven't touched `tbl_adverse_events()` yet.)

- I created an additional helper function `get_unique()` to retrieve unique values of a variable.
- I realized that in supplying extra ids to get base n, we also need to know which strata the ids belong to. I modified the input ids argument `id_df` to take a frame of ids & strata.
- I couldn't get the cute `complete` + `nesting` combo to work, had to resort to `expand_grid` + `right_join`.

I'm open to any ideas or improvements, let me know what you think!

**Review by ddsjoberg — CHANGES_REQUESTED** [review]

**Comment by ddsjoberg** [comment]
This example is from the unit tests. I don't think the complete data has all the needed observations. For example, ID1 needs a dummy observation for the bruising AE, right?

``` r
library(gtreg)

df1 <- tibble::tibble(
  patient_id = paste0("ID", c(1,1,2,3)),
  system_organ_class = "Blood and lymphatic system disorders",
  adverse_event = c("Anaemia", "Anaemia", "Anaemia", "Increased tendency to bruise"),
  grade = c(1, 1, 1, 2),
  trt = rep("A", 4)
)

e1 <- complete_data(
  data = df1,
  id = patient_id,
  ae = adverse_event,
  soc = system_organ_class,
  by = grade,
  strata = trt,
  id_df = data.frame(
    patient_id = paste0("ID", 1:5),
    trt = c(rep("A", 3),rep("B", 2))
  ),
  by_values = c("Unknown", 1:3)
)

df1
#> # A tibble: 4 x 5
#>   patient_id system_organ_class                   adverse_event      grade trt  
#>   <chr>      <chr>                                <chr>              <dbl> <chr>
#> 1 ID1        Blood and lymphatic system disorders Anaemia                1 A    
#> 2 ID1        Blood and lymphatic system disorders Anaemia                1 A    
#> 3 ID2        Blood and lymphatic system disorders Anaemia                1 A    
#> 4 ID3        Blood and lymphatic system disorders Increased tendenc~     2 A
e1$data_complete
#> # A tibble: 5 x 6
#>   id    strata ae                           soc                by    in_original
#>   <chr> <chr>  <chr>                        <chr>              <fct> <lgl>      
#> 1 ID1   A      Anaemia                      Blood and lymphat~ 1     TRUE       
#> 2 ID2   A      Anaemia                      Blood and lymphat~ 1     TRUE       
#> 3 ID3   A      Increased tendency to bruise Blood and lymphat~ 2     TRUE       
#> 4 ID4   B      Increased tendency to bruise Blood and lymphat~ dummy NA         
#> 5 ID5   B      Increased tendency to bruise Blood and lymphat~ dummy NA
```

<sup>Created on 2021-12-14 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

File: R/get_unique.R [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    dplyr::distinct( {{var}} ) %>%
+    dplyr::arrange( {{var}} ) %>%
+    {if ( drop_na ) tidyr::drop_na(., {{var}} ) else . } %>%
+    dplyr::pull( {{var}} )
+
+  if (class(values) == "factor") values <- as.character(values)
+
+  return(values)
+}
+
```
Comment (tgerke) [inline R/get_unique.R:36]: I would hoist a return like this
```suggestion

  if (class(values) == "factor") {
    values <- as.character(values)
    return(values)
  }
  
  data %>%
    dplyr::distinct( {{var}} ) %>%
    dplyr::arrange( {{var}} ) %>%
    {if ( drop_na ) tidyr::drop_na(., {{var}} ) else . } %>%
    dplyr::pull( {{var}} )
  
}
```

File: DESCRIPTION [hunk 2]
```diff
@@ -11,6 +11,7 @@ Depends:
     R (>= 2.10)
 Imports: 
     broom.helpers (>= 1.4.0),
+    forcats,
```
[the remark this answers is not in the document]
Comment (shannonpileggi) [reply DESCRIPTION]: thanks for the tip! added in commit to this PR

File: R/complete_data.R [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+#' @param id_df Optional data frame of complete id values and strata to achieve correct
+#' base n for the situation in which not all subjects experience adverse events
+#' @param by_values Optional vector of complete by values, listed in desired order,
+#' to achieve correct table structure for the situation in which an adverse
+#' event of a certain grade is not observed for a given soc
+#'
+#'
+#' @export
+#' @examples
+#' df_aes %>%
```
[the remark this answers is not in the document]
Comment (shannonpileggi) [reply R/complete_data.R]: I believe this was resolved in your PR; we can add more examples as well. Did we decide that user is to pass quoted strings to argument rather than unquoted variable names?

File: R/complete_data.R [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+#'   tbl_adverse_events(
+#'     id = patient_id,
+#'     ae = adverse_event,
+#'     soc = system_organ_class,
+#'     by = grade,
+#'     strata = trt
+#'   ) %>%
+#'   as_kable() # UPDATE THIS WITH PROPER gt image at some point..
+
+complete_data <- function(data, id, ae, soc, by, strata,
```
[the remark this answers is not in the document]
Comment (shannonpileggi) [reply R/complete_data.R]: completed with your PR

File: R/complete_data.R [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+  lst_name_recode <- list(id = id, strata = strata,
+                          ae = ae, soc = soc, by = by)
+
+  # some default factor levels -------------------------------------------------
+  initial_missing <- "Not entered"
+  initial_dummy   <- "dummy"
+
+  # initial data renaming and trimming -----------------------------------------
+  data_initial <- data
+    dplyr::rename(!!!lst_name_recode) %>%
```
[the remark this answers is not in the document]
Comment (shannonpileggi) [reply R/complete_data.R]: I think the renaming makes the programming easier, right? And if function is purely backstage it is safe to rename, if the function is user facing we could restore original names at end?

File: R/complete_data.R [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+  strata <-
+    .select_to_varnames({{ strata }}, data = data,
+                        arg_name = "strata", select_single = TRUE)
+
+  # list to rename variables----------------------------------------------------
+  lst_name_recode <- list(id = id, strata = strata,
+                          ae = ae, soc = soc, by = by)
+
+  # some default factor levels -------------------------------------------------
+  initial_missing <- "Not entered"
```
[the remark this answers is not in the document]
Comment (shannonpileggi) [reply R/complete_data.R]: agreed on both counts; both resolved with your PR

File: R/complete_data.R (hunk 3 above)
[the remark this answers is not in the document]
Comment (ddsjoberg) [reply R/complete_data.R]: I thought that because this function is used internally, we didn't need to use the selectors twice. But it's not much overhead at all to add that functionality back in!

File: R/complete_data.R (hunk 5 above)
[the remark this answers is not in the document]
Comment (ddsjoberg) [reply R/complete_data.R]: love it! sounds good to me!

File: R/complete_data.R (hunk 3 above)
[the remark this answers is not in the document]
Comment (shannonpileggi) [reply R/complete_data.R]: Sounds good!  Let's chat order of operations and when we expect `.complete_ae_data()` to be called.

File: R/get_unique.R [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+#' Values are returned in their respective natural ordering, whether it be numeric,
+#' character, or factor. If the input variable is a factor, get_unique()
+#' removes factor levels and returns a character vector.
+#'
+#' @param data data frame
+#' @param var variable from which we want unique values
+#' @param drop_na whether or not to drop NA from returned values; default to TRUE
+#'
+#' @return a vector (either character or numeric, depending on input)
+#' @export
```
[the remark this answers is not in the document]
Comment (shannonpileggi) [reply R/get_unique.R]: ok i added this. never used it before and could use another eye to make sure it is as intended!

File: R/get_unique.R [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+#' dat <- data.frame(
+#'   x = c(1, 1, 2),
+#'   y = c("B", "B", "A"),
+#'   z = factor(x = c("cat", "cat", "dog"), levels = c("dog", "cat"))
+#'   )
+#' get_unique(dat, x)
+#' get_unique(dat, y)
+#' get_unique(dat, z)
+#'
+get_unique <- function(data, var, drop_na = TRUE){
```
[the remark this answers is not in the document]
Comment (shannonpileggi) [reply R/get_unique.R]: agreed.  you might have some slicker solutions, but this functionality has been added in.

**Comment by shannonpileggi** [comment]
Not sure how to add this as a comment in the review, but I think we could leave `by_values` flexible to input either character or numeric as they eventually get coerced to factor anyway.  Thoughts?

**Comment by shannonpileggi** [comment]
What is the `, NA` doing in the testing?

**Comment by shannonpileggi** [comment]
Just wanted to say purrr::compact() is cool!

File: R/complete_data.R [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+  # combine dummy and mising with by values ------------------------------------
+  by_values <- c(initial_dummy, initial_missing, by_values)
+
+
+  # retrieve unique values for ae and soc --------------------------------------
+  soc_values <- get_unique(data_initial, soc)
+  ae_values  <- get_unique(data_initial, ae)
+
+  # if data frame of ids is supplied -------------------------------------------
+  if (!is.null(id_df)) {
```
[the remark this answers is not in the document]
Comment (shannonpileggi) [reply R/complete_data.R]: done

File: R/complete_data.R [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+                          ae = ae, soc = soc, by = by)
+
+  # some default factor levels -------------------------------------------------
+  initial_missing <- "Not entered"
+  initial_dummy   <- "dummy"
+
+  # initial data renaming and trimming -----------------------------------------
+  data_initial <- data
+    dplyr::rename(!!!lst_name_recode) %>%
+    dplyr::select(all_of(names(lst_name_recode)))
```
[the remark this answers is not in the document]
Comment (shannonpileggi) [reply R/complete_data.R]: thanks for flagging!

File: R/complete_data.R [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+  # if data frame of ids is not supplied  --------------------------------------
+  if (is.null(id_df)) {
+    id_df <- data_initial
+    }
+
+  # retrieve unique strata & id combinations -----------------------------------
+  id_df <- id_df %>%
+      dplyr::select(id, strata) %>%
+      dplyr::distinct(id, strata) %>%
+      dplyr::arrange(strata, id)
```
[the remark this answers is not in the document]
Comment (shannonpileggi) [reply R/complete_data.R]: i think this bit was re-factored and is no longer needed

File: R/complete_data.R [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+
+  # fully expanded data frame --------------------------------------------------
+  data_full <- id_df %>%
+    tidyr::expand_grid(
+      soc = soc_values,
+      ae = ae_values
+      )
```
[the remark this answers is not in the document]
Comment (shannonpileggi) [reply R/complete_data.R]: agreed we need a bajillion tests, but i think this specific bit of code has been refactored so resolving

**Comment by ddsjoberg** [comment]
> Not sure how to add this as a comment in the review, but I think we could leave `by_values` flexible to input either character or numeric as they eventually get coerced to factor anyway. Thoughts?

Yeah, we can make that work!

**Comment by ddsjoberg** [comment]
> What is the `, NA` doing in the testing? Does it mean expecting no errors?

exactly :)

File: R/complete_data.R [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+      ),
+      # convert by variable to factor ------------------------------------------
+      by = forcats::as_factor(by),
+      by = forcats::fct_expand(by, by_values),
+      by = forcats::fct_relevel(by, by_values)
+    ) %>%
+    dplyr::arrange(.data$id, .data$soc,  .data$by) %>%
+    # keep highest `grade` value per patient, e.g. highest grade
+    dplyr::group_by(.data$id, .data$soc) %>%
+    dplyr::slice_tail(n = 1) %>%
```
[the remark this answers is not in the document]
Comment (shannonpileggi) [reply R/complete_data.R]: that makes sense, and i believe this was addressed with the PR. i added in `tab_vars` vector b/c i thought it might make it simpler, but i think it is making it harder, so i am totally open to rolling back that commit.

File: R/complete_data.R [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+    dplyr::ungroup()
+
+  # will return inputs ---------------------------------------------------------
+  # keep all of these or just some?
+  tbl_ae_inputs <- as.list(environment())
```
[the remark this answers is not in the document]
Comment (shannonpileggi) [reply R/complete_data.R]: let's keep discussing where and how `complete_ae_data` will be used to make sure we get this right

File: R/complete_data.R [hunk 15]
```diff
@@ -0,0 +1,137 @@
+#' Create a complete and expanded data frame for tabulating adverse events
+#'
+#' @param data Data frame
+#' @param id Variable name of the patient ID
+#' @param soc Variable name of the system organ class column
+#' @param ae Variable name of the adverse event column
+#' @param by Variable to split results by, e.g. report AEs by grade or attribution
```
[the remark this answers is not in the document]
Comment (shannonpileggi) [reply R/complete_data.R]: probably - i think you addressed this with PR

**Comment by shannonpileggi** [comment]
> This example is from the unit tests. I don't think the complete data has all the needed observations. For example, ID1 needs a dummy observation for the bruising AE, right?
> 
> ```r
> library(gtreg)
> 
> df1 <- tibble::tibble(
>   patient_id = paste0("ID", c(1,1,2,3)),
>   system_organ_class = "Blood and lymphatic system disorders",
>   adverse_event = c("Anaemia", "Anaemia", "Anaemia", "Increased tendency to bruise"),
>   grade = c(1, 1, 1, 2),
>   trt = rep("A", 4)
> )
> 
> e1 <- complete_data(
>   data = df1,
>   id = patient_id,
>   ae = adverse_event,
>   soc = system_organ_class,
>   by = grade,
>   strata = trt,
>   id_df = data.frame(
>     patient_id = paste0("ID", 1:5),
>     trt = c(rep("A", 3),rep("B", 2))
>   ),
>   by_values = c("Unknown", 1:3)
> )
> 
> df1
> #> # A tibble: 4 x 5
> #>   patient_id system_organ_class                   adverse_event      grade trt  
> #>   <chr>      <chr>                                <chr>              <dbl> <chr>
> #> 1 ID1        Blood and lymphatic system disorders Anaemia                1 A    
> #> 2 ID1        Blood and lymphatic system disorders Anaemia                1 A    
> #> 3 ID2        Blood and lymphatic system disorders Anaemia                1 A    
> #> 4 ID3        Blood and lymphatic system disorders Increased tendenc~     2 A
> e1$data_complete
> #> # A tibble: 5 x 6
> #>   id    strata ae                           soc                by    in_original
> #>   <chr> <chr>  <chr>                        <chr>              <fct> <lgl>      
> #> 1 ID1   A      Anaemia                      Blood and lymphat~ 1     TRUE       
> #> 2 ID2   A      Anaemia                      Blood and lymphat~ 1     TRUE       
> #> 3 ID3   A      Increased tendency to bruise Blood and lymphat~ 2     TRUE       
> #> 4 ID4   B      Increased tendency to bruise Blood and lymphat~ dummy NA         
> #> 5 ID5   B      Increased tendency to bruise Blood and lymphat~ dummy NA
> ```
> 
> Created on 2021-12-14 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)

Agreed - I did not fully consider the input required by the tabulate function.

File: R/get_unique.R (hunk 1 above)
Comment (shannonpileggi) [reply R/get_unique.R]: This function has been refactored slightly.

File: R/complete_data.R [hunk 16]
```diff
@@ -0,0 +1,137 @@
...
+#'     by = grade,
+#'     strata = trt
+#'   ) %>%
+#'   as_kable() # UPDATE THIS WITH PROPER gt image at some point..
+
+complete_data <- function(data, id, ae, soc, by, strata,
+                          id_df, by_values) {
+
+  # evaluate bare selectors/check inputs ---------------------------------------
+  stopifnot(inherits(data, "data.frame"))
```
[the remark this answers is not in the document]
Comment (shannonpileggi) [reply R/complete_data.R]: was this check removed? add back in?

**Review by shannonpileggi — COMMENTED** [review]
@ddsjoberg this bit might have been more trouble than it is worth - not sure. feel free to roll back.

if `ae` and `id` are the only required arguments, then we just option here for `soc` and `by`, as `strata` is implicitly tied to `id`?

**Comment by shannonpileggi** [comment]
Is this PR getting hard to follow for anyone else? Would it be easier to merge and start new PR for fresh eyes?

File: R/complete_data.R (hunk 16 above)
[the remark this answers is not in the document]
Comment (ddsjoberg) [reply R/complete_data.R]: I think i removed it because it's an internal function, and the check will already appear in the `tbl_*()` function. We can add it back, since I do think we should export this function.

File: R/complete_data.R (hunk 13 above)
[the remark this answers is not in the document]
Comment (ddsjoberg) [reply R/complete_data.R]: I took a quick look at the `tab_vars`. I think there are two potential issues.
1. The `arrange()` and `group_by()` are both by `tab_vars`, but the `group_by()` cannot include `"by"`, as we are selecting the AE with the highest grade (ie most severe within by variable).
2. I saw that `..soc..` was removed. That column was meant to indicate the row that should be summarized when summarizing the SOC, which is not the same row as the AE summary. How do you envision identifying the appropriate row?

**Comment by ddsjoberg** [comment]
> Is this PR getting hard to follow for anyone else? Would it be easier to merge and start new PR for fresh eyes?

YES! Can we sort out the tab_vars, then merge? 


> @ddsjoberg this bit might have been more trouble than it is worth - not sure. feel free to roll back.
> 
> if `ae` and `id` are the only required arguments, then we just option here for `soc` and `by`, as `strata` is implicitly tied to `id`?

I am not sure if I follow 100%? Let's revisit in the fresh PR?

**Comment by shannonpileggi** [comment]
> > Is this PR getting hard to follow for anyone else? Would it be easier to merge and start new PR for fresh eyes?
> 
> YES! Can we sort out the tab_vars, then merge?
> 
> > @ddsjoberg this bit might have been more trouble than it is worth - not sure. feel free to roll back.
> > if `ae` and `id` are the only required arguments, then we just option here for `soc` and `by`, as `strata` is implicitly tied to `id`?
> 
> I am not sure if I follow 100%? Let's revisit in the fresh PR?

Agree, let's drop tab_vars and merg and fresh PR.  I'm getting into git trouble over here. I tried to roll back to previous commit before I added that but I don't think it worked.

**Review by ddsjoberg — APPROVED** [review]
